Add "Synesthesia" mod for osu! ruleset

jtbiddle opened the pull request:
This is a mod for osu!standard that colors hit circles and sliders based on their beat division in the song. For example, objects that are on full notes are white, half notes are red, and quarter notes are blue. This is a simple implementation using the function for the snap coloring preference in osu!mania, and shares the same colors. I implemented this in an attempt to make reading complicated rhythms easier, especially for low AR or hidden. 

https://github.com/ppy/osu/assets/1653385/ce1dd4e8-d012-4be9-a581-6a01f71d65c1


Compatibility with other mods has been tested.  I could imagine this being combined with the "Incognito" mod which also changes hit object colors ( #22271) to avoid cluttering the mod menu.

jtbiddle committed:
Initial commit for the snap colour mod. Implements basic functionality.

jtbiddle committed:
Formatting fixes

ItsShamed commented on osu.Game.Rulesets.Osu/Mods/OsuModSnapColour.cs:
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
+
```
No `#nullable disable` in new code
https://github.com/ppy/osu/wiki/Nullable-Reference-Types-(NRT)#dependency-injection-considerations

jtbiddle replied:
Thank you very much for the quick code review, I removed #nullable disable and made your suggested fixes

ItsShamed commented on osu.Game/Rulesets/Mods/ModSnapColour.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        public override string Name => "Snap Colour";
+        public override string Acronym => "SC";
+        public override LocalisableString Description => new LocalisableString("Colours hit objects based on the rhythm.");
```
If you want to make this string localisable this is not the way to do it, please look at how it's done in the `osu.Game.Localisation` namespace/folder. Otherwise, you can just use a primitive string, there is an implicit type conversion, you'll be fine.

jtbiddle replied:
Thank you, I just changed it to a primitive string for now, can always make it (properly) localisable later if needed

ItsShamed replied:
Localisation for mod descriptions should probably be done on https://github.com/ppy/osu-web rather than here

ItsShamed commented on osu.Game.Rulesets.Osu/Mods/OsuModSnapColour.cs:
```diff
@@ -0,0 +1,45 @@
+    /// </summary>
+    public class OsuModSnapColour : ModSnapColour, IApplicableToBeatmap, IApplicableToDrawableHitObject
+    {
+        [Resolved]
+        private OsuColour colours { get; set; } = new OsuColour();
```
This is not how Dependency Injection works. You don't need to initialise the auto-property, with NRT enabled use the null-forgiving operator (see https://github.com/ppy/osu/wiki/Nullable-Reference-Types-(NRT)#dependency-injection-considerations)

jtbiddle replied:
Thanks, I removed the dependency injection and initialized OsuColour as readonly. 

I could be wrong but it seems like the OsuColour dependency injection was implemented for Drawable classes so it needed to be initialized in this class

ItsShamed replied:
Yes, but you don't need to initialise the auto-property in such cases. You only need to initialise it to `null! because of NRT`. Dependencies will be provided automatically when the drawable will be loaded.

peppy commented:
It's a touch weird that this is introduced as a mod while it exists as a setting in osu!mania. Not necessarily the wrong approach, but food for thought.

ItsShamed commented on osu.Game.Rulesets.Osu/Mods/OsuModSnapColour.cs:
```diff
@@ -0,0 +1,45 @@
+        [Resolved]
+        private OsuColour colours { get; set; } = new OsuColour();
+
+        [Resolved(canBeNull: true)]
+        private IBeatmap currentBeatmap { get; set; }
```
```suggestion
        [Resolved]
        private IBeatmap? currentBeatmap { get; set; }
```

ItsShamed reviewed (changes requested):
Quick code quality check

jtbiddle committed:
Code quality improvements (thanks to ItsShamed): Removed #nullable disable, fixed incorrect LocalisableString, removed incorrect dependency injection for OsuColour, fixed nullable dependency for IBeatmap

Removed unnecessary usage of "this." caught by the CI code quality check

bdach commented:
> It's a touch weird that this is introduced as a mod while it exists as a setting in osu!mania.

I really think this is the big issue here and one that doesn't really warrant per-ruleset considerations. The way I see it, either toggling the colour snapping is visual preference and as such should be a game setting, or it actually changes the difficulty of gameplay, and as such should be a mod. I'm not sure it should be allowed to be a setting in one ruleset but a mod in the other.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModSnapColour.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+            if (currentBeatmap.IsNull() || drawable.IsNull()) return;
+
+            drawable.OnUpdate += _ =>
+                drawable.AccentColour.Value = BindableBeatDivisor.GetColourFor(
+                    currentBeatmap.ControlPointInfo.GetClosestBeatDivisor(drawable.HitObject.StartTime),
+                    colours);
```
This should not be running every frame. `DrawableHitObject.ApplyCustomUpdateState` was intended for this purpose and is likely to be equivalent in functionality (but vastly better in performance).

jtbiddle replied:
Thanks, I have changed it to use ApplyCustomUpdateState.

(It was originally written using that, but it also requires disabling the combo color update in DrawableNote to work and I was trying to touch other code as little as possible. This is a better way to do it for performance, though.)

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModSnapColour.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        private readonly OsuColour colours = new OsuColour();
+
+        [Resolved]
+        private IBeatmap? currentBeatmap { get; set; }
```
This `[Resolved]` does nothing. As you [correctly pointed out](https://github.com/ppy/osu/pull/23520#discussion_r1192926111), this dependency injection mechanism only works for drawables. So the attribute should be removed as it's misleading at best.

jtbiddle replied:
Thanks, I have removed the resolved tags

bdach reviewed (changes requested):
Additionally:

peppy commented:
> I really think this is the big issue here and one that doesn't really warrant per-ruleset considerations. The way I see it, either toggling the colour snapping is visual preference and as such should be a game setting, or it actually changes the difficulty of gameplay, and as such should be a mod. I'm not sure it should be allowed to be a setting in one ruleset but a mod in the other.

From that angle it could be argued that the mania setting is moved to exist as a mod. It's the kind of thing I think *should* be shared in replays, so that might be a good direction?

jtbiddle committed:
Removed unnecessary [Resolved] attribute (thanks bdach)
Moved accent color assignment from OnUpdate to ApplyCustomUpdateState. In order to get this to work, a flag needed to be added to DrawableHitObject.cs to disable combo color updates also being applied.

jtbiddle committed:
Removed import for Allocation and other unnecessary [Resolved] tag missed in last commit

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -74,6 +74,11 @@ public abstract partial class DrawableHitObject : PoolableDrawableWithLifetime<H
+        /// <summary>
+        /// Whether this object should be coloured using its combo position
+        /// </summary>
+        public bool EnableComboColour { get; set; } = true;
```
I'd like to say this flag shouldn't be required to make this work... What is the reason for adding it?

jtbiddle replied:
When testing, for some reason the snap coloring would only apply to the first few notes of a song. After that, it would revert to the original combo colors with occasional off-colors. Here are two captures I took without the flag, the first showing the snap color ending, and the second showing off-colors after the snap coloring ends:


https://github.com/ppy/osu/assets/1653385/393255b2-1dcd-487b-9d56-bb26713065f8


https://github.com/ppy/osu/assets/1653385/b190d2cc-d8e2-4167-ade3-11785da64eec

jtbiddle replied:
I looked into the issue more, and I think I narrowed it down: There's an additional call to UpdateComboColour at the end of OnApply in DrawableHitObject on line 283 with the following comment:
``` 
// Combo colour may have been applied via a bindable flow while no object entry was attached.
// Update here to ensure we're in a good state.
UpdateComboColour();
```

This is only called after the first few objects in the map. All objects, including the first objects in the map, have UpdateComboColour called from the bindings in LoadComplete around line 205:
```
comboIndexBindable.BindValueChanged(_ => UpdateComboColour());
comboIndexWithOffsetsBindable.BindValueChanged(_ => UpdateComboColour(), true);
```

Only these calls are overridden by changing the color in ApplyCustomUpdateState. If I comment out the UpdateComboColour() on line 283 above, the snap coloring works for the whole map without the flag added (but I have yet do do enough testing to see what other side effects this has.)

bdach replied:
Thanks for the explanation.

I tried to see if this is fixable by overriding the combo colour on `drawableObject.HitObjectApplied` too, but unfortunately it doesn't do the trick (rewinding replays, or changing combo colour normalisation break that). It appears that the combo colour is too deeply coupled into everything else to be changed reliably via existing hooks.

We aim for maximal separation of mods from game logic, so I would like to see this flag gone.

Possible ways out to explore:
- Revert back to overriding the colour in `Update()`
- Add some special interface for this goal that can be used to hook into the `UpdateComboColour()` flow (`IApplicableToComboColours`? `IApplicableToSkin` maybe, even)?

Due to the number of other open pulls and/or lack of time I am not able to put time in to test these approaches out. Maybe we could be fine with the crude `Update()` thing for now. Dunno.

peppy replied:
I'd be okay with `Update` as a short term solution. Beats having one more PR open indefinitely.

jtbiddle replied:
I changed it back to using OnUpdate and removed the changes to DrawableHitObject for now.

jtbiddle committed:
Reverted to applying the color change in OnUpdate, removed EnableComboColour flag from DrawableHitObject.cs

jtbiddle commented:
Thanks for the reviews so far, let me know if there are any changes you would like to see to it. I wanted to keep this PR simple and to see if there was interest in this addition. 

Some ideas for this mod that I would be happy to refine with the team and implement in the future:
- Moving it to a preference instead of a mod if that is more appropriate
- Color customization, possibly with an option to read colors from the current skin. (Fixing #23684 would make adding a color picker to the mod customization menu much easier)
- Either disabling or implementing the "Combo Colour Normalisation" and "Beatmap Colours" options when the mod is selected
- Allowing beatmap colors to be used if the number of different time divisions in a beatmap is equal to or less than the number of beatmap colors (would require pre-processing the beatmap on loading)
- A toggle to disable slider tail coloring. It currently provides more timing information as it is, but may look nicer if sliders are a uniform color

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModSnapColour.cs:
```diff
@@ -0,0 +1,40 @@
+
+        public void ApplyToDrawableHitObject(DrawableHitObject drawable)
+        {
+            if (currentBeatmap.IsNull() || drawable.IsNull()) return;
```
This is strange for two reasons:

- Why use `.IsNull()` over a plain null check?
- `drawable` should not be null, ever.

This should be

```suggestion
            if (currentBeatmap == null) return;
```

if you ask me.

bdach commented on osu.Game/Rulesets/Mods/ModSnapColour.cs:
```diff
@@ -0,0 +1,19 @@
+    /// </summary>
+    public class ModSnapColour : Mod
+    {
+        public override string Name => "Snap Colour";
+        public override string Acronym => "SC";
```
Not sure on the naming of this thing. The existing mania setting is named "timing-based colouring". I would like to keep that naming consistent, but "timing-based colouring" a bit too much of a mouthful to use for a mod name. "Snap" is a bit of an editor term and I'm not sure it should be used in the mod name.

Maybe something like "Coloured Rhythm" could work? Or - going off the wall a bit - ["Synesthesia"](https://en.wikipedia.org/wiki/Synesthesia)?

jtbiddle replied:
Yeah, the name was just taken from the function name - I figured that was some "standard" name for it in VSRGs but I'm not tied to it at all. I really like the suggestion of "Synesthesia", a cool reference and in-line with the other mod names. I've renamed it to Synesthesia in the latest commit.

bdach commented on osu.Game/Rulesets/Mods/ModSnapColour.cs:
```diff
@@ -0,0 +1,19 @@
+        public override string Acronym => "SC";
+        public override LocalisableString Description => "Colours hit objects based on the rhythm.";
+        public override double ScoreMultiplier => 1;
+        public override ModType Type => ModType.Conversion;
```
I'm not sure why this ended up in `ModType.Conversion` but I don't think the mod matches that. `ModType.Fun` is where I'd put it.

bdach reviewed (commented):
Let's keep away from the scope creep mentioned above and focus on what we have right now - still a few things I'm not sure about

jtbiddle committed:
Added recommendations from bdach:
Fixed null checking in ApplyToDrawableHitObject
Renamed mod to "Synesthesia"
Moved to the "Fun" mod category

bdach commented:
I don't have any problems with this anymore, but I also don't feel confident enough to proceed with this mod as it is written on the basis of my judgement alone. Going to need a second approval for this to go forward.

peppy committed:
Merge branch 'master' into snap-colour-mod

peppy committed:
Avoid doing expensive colour fetch operation every update

peppy commented:
@bdach please double-check my changes. I wanted to avoid the expensive colour lookup (including a `TimingPointAt`) multiple times per frame.

bdach committed:
Fix slider tails receiving wrong colours

Only visually apparent on legacy skins.

bdach commented:
Fixed one more issue with slider tails that was only visible on legacy skins ([see discord conversation](https://discord.com/channels/188630481301012481/188630652340404224/1120372851753631744)). As stated there, not sure what the reaction to differently-coloured slider pieces is going to be, but let's try this and see what it will be.
